File: internal/capability/audit.go
package capability

import (
	"context"
	"fmt"
	"net/url"
	"strconv"
	"strings"
	"time"

	"github.com/ro-ag/ptrack/internal/model"
)

const globalAuditLimit = 5000

// AuditStore is the narrow persistence contract used by Recorder.
type AuditStore interface {
	AddCapabilityAuditBounded(model.CapabilityAudit, int, int) (model.CapabilityAudit, error)
}

// AuditEvent contains transient operation metadata. Target is sanitized by
// capability kind before persistence and ErrorClass is reduced to a fixed set.
type AuditEvent struct {
	Operation     string
	Target        string
	Success       bool
	ErrorClass    string
	Duration      time.Duration
	RequestBytes  int64
	ResponseBytes int64
	Redirects     int
}

// Recorder persists bounded metadata only. It deliberately has no fields for
// headers, bodies, terminal output, credentials, raw stderr, or arguments.
type Recorder struct {
	Store AuditStore
	Now   func() time.Time
}

// Record sanitizes and appends one event when the capability's audit policy is
// enabled. A nil Store is treated as an unavailable optional audit sink.
func (r Recorder) Record(_ context.Context, capability model.Capability, event AuditEvent) error {
	if !capability.Audit.Enabled || r.Store == nil {
		return nil
	}
	now := time.Now
	if r.Now != nil {
		now = r.Now
	}
	record := model.CapabilityAudit{
		CapabilityID:   capability.ID,
		AgentProfile:   sanitizeProfile(capability.AgentProfile),
		Kind:           capability.Kind,
		Operation:      sanitizeOperation(capability.Kind, event.Operation),
		Target:         sanitizeAuditTarget(capability.Kind, event.Target),
		Success:        event.Success,
		ErrorClass:     sanitizeErrorClass(event.Success, event.ErrorClass),
		DurationMillis: boundedDurationMillis(event.Duration),
		RequestBytes:   boundedCount(event.RequestBytes),
		ResponseBytes:  boundedCount(event.ResponseBytes),
		Redirects:      boundedRedirects(event.Redirects),
		CreatedAt:      now(),
	}
	_, err := r.Store.AddCapabilityAuditBounded(record, capability.Audit.RetainLast, globalAuditLimit)
	return err
}

var allowedErrorClasses = map[string]bool{
	"none": true, "denied": true, "dns": true, "routing": true, "vpn": true,
	"proxy": true, "tls": true, "host-key": true, "authentication": true,
	"sandbox": true, "remote-policy": true, "timeout": true, "transport": true,
	"request-limit": true, "response-limit": true, "output-limit": true,
	"cancelled": true, "internal": true,
}

func sanitizeErrorClass(success bool, value string) string {
	if success {
		return "none"
	}
	value = strings.ToLower(strings.TrimSpace(value))
	if allowedErrorClasses[value] {
		return value
	}
	return "internal"
}

func sanitizeOperation(kind model.CapabilityKind, value string) string {
	value = strings.ToLower(strings.TrimSpace(value))
	allowed := false
	switch kind {
	case model.CapabilityHTTP:
		allowed = allowedHTTPMethods[strings.ToUpper(value)] || value == "test"
	case model.CapabilityGit:
		allowed = contains([]string{"status", "fetch", "pull", "push", "ls-remote", "test"}, value)
	case model.CapabilitySSH:
		allowed = contains([]string{
			string(SSHGit), string(SSHRemoteCommand), string(SSHUpload), string(SSHDownload),
			string(SSHInteractiveShell), string(SSHLocalForward), string(SSHRemoteForward), "test",
		}, value)
	}
	if !allowed {
		return "unknown"
	}
	return value
}

func sanitizeAuditTarget(kind model.CapabilityKind, raw string) string {
	switch kind {
	case model.CapabilityHTTP:
		u, err := url.Parse(raw)
		if err != nil || u.Scheme == "" || u.Host == "" {
			return "invalid-http-target"
		}
		// Paths are caller-controlled and commonly carry opaque IDs or
		// credentials. Persist only the approved network origin.
		return truncate(u.Scheme+"://"+u.Host, 256)
	case model.CapabilityGit:
		name, err := normalizeToken(raw, "Git remote")
		if err != nil {
			return "invalid-git-target"
		}
		return truncate("remote:"+name, 160)
	case model.CapabilitySSH:
		host, port, err := netSplitAuditTarget(raw)
		if err != nil {
			return "invalid-ssh-target"
		}
		return truncate(host+":"+strconv.Itoa(port), 256)
	default:
		return "unknown-target"
	}
}

func netSplitAuditTarget(raw string) (string, int, error) {
	value := strings.TrimSpace(raw)
	colon := strings.LastIndex(value, ":")
	if colon <= 0 {
		return "", 0, fmt.Errorf("missing port")
	}
	host, err := normalizeHost(strings.Trim(value[:colon], "[]"))
	if err != nil {
		return "", 0, err
	}
	port, err := strconv.Atoi(value[colon+1:])
	if err != nil || port < 1 || port > 65535 {
		return "", 0, fmt.Errorf("invalid port")
	}
	return host, port, nil
}

func sanitizeProfile(value string) string {
	profile, err := normalizeProfile(value)
	if err != nil {
		return "unknown-profile"
	}
	return profile
}

func boundedDurationMillis(value time.Duration) int64 {
	if value < 0 {
		return 0
	}
	maximum := 24 * time.Hour
	if value > maximum {
		value = maximum
	}
	return value.Milliseconds()
}

func boundedCount(value int64) int64 {
	if value < 0 {
		return 0
	}
	const maximum = int64(1 << 40)
	if value > maximum {
		return maximum
	}
	return value
}

func boundedRedirects(value int) int {
	if value < 0 {
		return 0
	}
	if value > maxRedirects {
		return maxRedirects
	}
	return value
}

func truncate(value string, maximum int) string {
	if len(value) <= maximum {
		return value
	}
	return value[:maximum]
}

File: crates/ptrack-capability-policy/src/audit.rs
use ptrack_core::{Capability, CapabilityAudit, CapabilityKind, Timestamp};

use crate::normalize::{
    normalize_endpoint, normalize_http_url, normalize_profile, normalize_token,
};

const MAX_DURATION_MILLIS: i64 = 24 * 60 * 60 * 1_000;
const MAX_BYTES: i64 = 1 << 40;
const MAX_REDIRECTS: i64 = 10;

/// Transient operation metadata. It deliberately cannot contain headers,
/// bodies, process output, credentials, raw stderr, or argv.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct AuditEvent {
    pub operation: String,
    pub target: String,
    pub success: bool,
    pub error_class: String,
    pub duration_millis: i64,
    pub request_bytes: i64,
    pub response_bytes: i64,
    pub redirects: i64,
}

/// Opaque sanitized metadata accepted by the project store.
#[derive(Clone, Debug)]
pub struct SanitizedAudit {
    record: CapabilityAudit,
    retain_last: i64,
}

/// Reduces a transient event to fixed, bounded metadata.
///
/// Returns `None` when auditing is disabled for the capability.
#[must_use]
pub fn sanitize_audit(capability: &Capability, event: &AuditEvent) -> Option<SanitizedAudit> {
    if !capability.audit.enabled {
        return None;
    }
    Some(SanitizedAudit {
        record: CapabilityAudit {
            id: 0,
            capability_id: capability.id,
            agent_profile: normalize_profile(&capability.agent_profile)
                .unwrap_or_else(|_| "unknown-profile".to_owned()),
            kind: capability.kind,
            operation: sanitize_operation(capability.kind, &event.operation),
            target: sanitize_target(capability.kind, &event.target),
            success: event.success,
            error_class: sanitize_error_class(event.success, &event.error_class),
            duration_millis: event.duration_millis.clamp(0, MAX_DURATION_MILLIS),
            request_bytes: event.request_bytes.clamp(0, MAX_BYTES),
            response_bytes: event.response_bytes.clamp(0, MAX_BYTES),
            redirects: event.redirects.clamp(0, MAX_REDIRECTS),
            created_at: Timestamp::Zero,
        },
        retain_last: capability.audit.retain_last,
    })
}

impl SanitizedAudit {
    /// Supplies the storage-owned timestamp and consumes the opaque record.
    #[doc(hidden)]
    #[must_use]
    pub fn into_store_parts(mut self, now: Timestamp) -> (CapabilityAudit, i64) {
        self.record.created_at = now;
        (self.record, self.retain_last)
    }
}

fn sanitize_operation(kind: CapabilityKind, raw: &str) -> String {
    let value = raw.trim().to_lowercase();
    let allowed = match kind {
        CapabilityKind::Http => matches!(
            value.as_str(),
            "get" | "head" | "post" | "put" | "patch" | "delete" | "options" | "test"
        ),
        CapabilityKind::Git => matches!(
            value.as_str(),
            "status" | "fetch" | "pull" | "push" | "ls-remote" | "test"
        ),
        CapabilityKind::Ssh => matches!(
            value.as_str(),
            "git"
                | "remote-command"
                | "upload"
                | "download"
                | "interactive-shell"
                | "local-forward"
                | "remote-forward"
                | "test"
        ),
    };
    if allowed { value } else { "unknown".to_owned() }
}

fn sanitize_target(kind: CapabilityKind, raw: &str) -> String {
    match kind {
        CapabilityKind::Http => normalize_http_url(raw, true).map_or_else(
            |_| "invalid-http-target".to_owned(),
            |url| truncate(format!("{}://{}", url.scheme, url.host), 256),
        ),
        CapabilityKind::Git => normalize_token(raw, "Git remote", true).map_or_else(
            |_| "invalid-git-target".to_owned(),
            |name| truncate(format!("remote:{name}"), 160),
        ),
        CapabilityKind::Ssh => normalize_endpoint(raw).map_or_else(
            |_| "invalid-ssh-target".to_owned(),
            |value| truncate(value, 256),
        ),
    }
}

fn sanitize_error_class(success: bool, raw: &str) -> String {
    if success {
        return "none".to_owned();
    }
    let value = raw.trim().to_lowercase();
    if matches!(
        value.as_str(),
        "denied"
            | "dns"
            | "routing"
            | "vpn"
            | "proxy"
            | "tls"
            | "host-key"
            | "authentication"
            | "sandbox"
            | "remote-policy"
            | "timeout"
            | "transport"
            | "request-limit"
            | "response-limit"
            | "output-limit"
            | "cancelled"
            | "internal"
    ) {
        value
    } else {
        "internal".to_owned()
    }
}

fn truncate(mut value: String, maximum: usize) -> String {
    if value.len() <= maximum {
        return value;
    }
    let mut end = maximum;
    while !value.is_char_boundary(end) {
        end -= 1;
    }
    value.truncate(end);
    value
}
